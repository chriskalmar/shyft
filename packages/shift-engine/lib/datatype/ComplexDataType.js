

class ComplexDataType {}


export default ComplexDataType


export const isComplexDataType = (obj) => {
  return (obj instanceof ComplexDataType)
}

