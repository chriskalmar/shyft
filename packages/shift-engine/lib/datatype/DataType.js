
import {
  passOrThrow,
  isFunction,
} from '../util';


class DataType {

  constructor (setup = {}) {

    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup

    passOrThrow(name, () => 'Missing data type name')
    passOrThrow(description, () => `Missing description for data type '${name}'`)

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`
    )

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`
      )

      this.validator = validate
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`
      )

      this.defaultValue = defaultValue
    }

    this.name = name
    this.description = description
    this.mock = mock

    if (enforceRequired) {
      this.enforceRequired = enforceRequired
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex
    }
  }


  validate = (value, context) => {
    if (value && this.validator) {
      this.validator(value, context)
    }
  }


  toString() {
    return this.name
  }

}


export default DataType


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}
