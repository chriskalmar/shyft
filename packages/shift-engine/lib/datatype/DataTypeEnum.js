
import {
  passOrThrow,
  isArray,
} from '../util';

import DataType from './DataType';


class DataTypeEnum extends DataType {

  constructor (setup = {}) {

    const {
      name,
      description,
      values,
    } = setup

    passOrThrow(
      isArray(values, true),
      () => `'Missing enum values for data type '${name}'`
    )

    super({
      ...setup,
      description: description || `Enumeration set: ${values.join(', ')}`,
      mock() {
        const randomPos = Math.floor( Math.random() * values.length )
        return values[ randomPos ]
      }
    })

    this.values = values
  }


  toString() {
    return this.name
  }

}


export default DataTypeEnum


export const isDataTypeEnum = (obj) => {
  return (obj instanceof DataTypeEnum)
}
