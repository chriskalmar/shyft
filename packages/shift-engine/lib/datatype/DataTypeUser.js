
import DataType from './DataType';


export default class DataTypeUser extends DataType {}

export const isDataTypeUser = (obj) => {
  return (obj instanceof DataTypeUser)
}
