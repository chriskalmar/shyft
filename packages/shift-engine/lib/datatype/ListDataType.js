
import {
  passOrThrow,
  isFunction,
} from '../util';

import { isEntity } from '../entity/Entity';
import { isDataType } from './DataType';
import ComplexDataType, { isComplexDataType } from './ComplexDataType';


class ListDataType extends ComplexDataType {

  constructor (setup = {}) {

    super()

    const {
      name,
      description,
      itemType,
    } = setup

    passOrThrow(name, () => 'Missing list data type name')
    passOrThrow(description, () => `Missing description for list data type '${name}'`)
    passOrThrow(itemType, () => `Missing item type for list data type '${name}'`)

    passOrThrow(
      isDataType(itemType) || isEntity(itemType) || isComplexDataType(itemType) || isFunction(itemType),
      () => `List data type '${name}' has invalid item type '${String(itemType)}'`
    )

    this.name = name
    this.description = description
    this.itemType = itemType
  }


  _processItemType() {
    return isFunction(this.itemType)
      ? this.itemType({
        name: this.name,
        description: this.description
      })
      : this.itemType
  }


  getItemType() {
    if (this._itemType) {
      return this._itemType
    }

    const ret = this._itemType = this._processItemType()
    return ret
  }


  toString() {
    return this.name
  }

}


export default ListDataType


export const isListDataType = (obj) => {
  return (obj instanceof ListDataType)
}


export const buildListDataType = (obj) => {
  return ({name, description}) => new ListDataType({
    description,
    ...obj,
    name,
  })
}
