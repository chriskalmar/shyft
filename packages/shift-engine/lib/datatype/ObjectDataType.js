
import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  isFunction,
} from '../util';

import { isEntity } from '../entity/Entity';
import { isDataType } from './DataType';
import ComplexDataType, { isComplexDataType } from './ComplexDataType';


class ObjectDataType extends ComplexDataType {

  constructor (setup = {}) {

    super()

    const {
      name,
      attributes,
    } = setup

    passOrThrow(name, () => 'Missing object data type name')
    passOrThrow(attributes, () => `Missing attributes for object data type '${name}'`)

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () => `Object data type '${name}' needs an attribute definition as a map or a function returning such a map`
    )

    this.name = name
    this._attributesMap = attributes
  }


  getAttributes () {
    if (this._attributes) {
      return this._attributes
    }

    const ret = this._attributes = this._processAttributeMap()
    return ret
  }


  _processAttribute (rawAttribute, attributeName) {

    if (rawAttribute instanceof ObjectDataType) {
      rawAttribute.getAttributes()
      return rawAttribute
    }

    const attribute = {
      ...rawAttribute,
      required: !!rawAttribute.required,
      name: attributeName
    }

    passOrThrow(attribute.description, () => `Missing description for '${this.name}.${attributeName}'`)

    passOrThrow(
      isDataType(attribute.type) || isEntity(attribute.type),
      () => `'${this.name}.${attributeName}' has invalid data type '${String(attribute.type)}'`
    )

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`
    )

    passOrThrow(
      !attribute.defaultValue || isFunction(attribute.defaultValue),
      () => `'${this.name}.${attributeName}' has an invalid defaultValue function'`
    )

    return attribute
  }


  _processAttributeMap () {

    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () => `Attribute definition function for object data type '${this.name}' does not return a map`
    )


    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Object data type '${this.name}' has no attributes defined`
    )

    const resultAttributes = {}

    attributeNames.forEach((attributeName) => {
      resultAttributes[ attributeName ] = this._processAttribute(attributeMap[ attributeName ], attributeName)
    })

    return resultAttributes
  }


  toString() {
    return this.name
  }

}


export default ObjectDataType


export const isObjectDataType = (obj) => {
  return (obj instanceof ObjectDataType)
}

