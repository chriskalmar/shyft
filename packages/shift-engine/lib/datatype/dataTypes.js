
import DataType from './DataType';
import DataTypeUser from './DataTypeUser';

import casual from 'casual';
import _ from 'lodash';


export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})



export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(-2^10, 2^10),
})


export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  mock: /* istanbul ignore next */
    () => casual.double(-2^10, 2^10),
})


export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  mock: /* istanbul ignore next */
    () => casual.boolean,
})


export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  mock: /* istanbul ignore next */
    () => casual.title,
})


export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  mock: /* istanbul ignore next */
    randomJson,
})


export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


/* istanbul ignore next */
function randomJson() {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase( casual.words(2) )
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer

    ret[ key ] = value
  })

  return ret
}
