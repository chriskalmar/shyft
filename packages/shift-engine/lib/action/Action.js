
import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  // isArray,
  isFunction,
} from '../util';

import { isDataType } from '../datatype/DataType';
import { isEntity } from '../entity/Entity';
import { isObjectDataType } from '../datatype/ObjectDataType';


class Action {

  constructor (setup = {}) {

    const {
      name,
      description,
      input,
      output,
      resolve,
    } = setup

    passOrThrow(name, () => 'Missing action name')
    passOrThrow(description, () => `Missing description for action '${name}'`)
    passOrThrow(input, () => `Missing input definition for action '${name}'`)
    passOrThrow(output, () => `Missing output definition for action '${name}'`)

    passOrThrow(
      isMap(input) || isFunction(input),
      () => `Action '${name}' needs an input definition as a map or a function returning such a map`
    )

    passOrThrow(
      isMap(output) || isFunction(output),
      () => `Action '${name}' needs an output definition as a map or a function returning such a map`
    )

    passOrThrow(
      isFunction(resolve),
      () => `Action '${name}' needs a resolve function`
    )

    this.name = name
    this.description = description
    this._inputMap = input
    this._outputMap = output
    this.resolve = resolve
  }



  getInput () {
    if (this._input) {
      return this._input
    }

    const ret = this._input = this._processParamMap(this._inputMap, true)
    return ret
  }


  hasInputParams () {
    return isMap(this.getInput(), true)
  }


  getOutput () {
    if (this._output) {
      return this._output
    }

    const ret = this._output = this._processParamMap(this._outputMap, false)
    return ret
  }


  hasOutputParams () {
    return isMap(this.getOutput(), true)
  }



  _processParam (rawParam, paramName) {

    if (isObjectDataType(rawParam)) {
      rawParam.getAttributes()
      return rawParam
    }

    const param = {
      ...rawParam,
      required: !!rawParam.required,
      name: paramName
    }

    passOrThrow(
      isDataType(param.type) || isEntity(param.type),
      () => `'${this.name}.${paramName}' has invalid data type '${String(param.type)}'`
    )

    passOrThrow(
      !param.defaultValue || isFunction(param.defaultValue),
      () => `'${this.name}.${paramName}' has an invalid defaultValue function'`
    )

    return param
  }


  _processParamMap (_paramMap, isInput=true) {

    // if it's a function, resolve it to get that map
    const paramMap = resolveFunctionMap(_paramMap);

    passOrThrow(
      isMap(paramMap),
      () => `${isInput ? 'Input' : 'Output'} definition function for action '${this.name}' does not return a map`
    )

    const paramNames = Object.keys(paramMap);
    const resultParams = {}

    paramNames.forEach((paramName) => {
      resultParams[ paramName ] = this._processParam(paramMap[ paramName ], paramName)
    })

    return resultParams
  }


  toString() {
    return this.name
  }

}


export default Action


export const isAction = (obj) => {
  return (obj instanceof Action)
}
