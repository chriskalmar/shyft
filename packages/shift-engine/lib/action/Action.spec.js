
import Action, {
  isAction,
} from './Action';

import { DataTypeString } from '../datatype/dataTypes';
import { buildObjectDataType } from '../datatype/ObjectDataType';

import {
  passOrThrow,
} from '../util';



describe('Action', () => {

  it('should have a name', () => {

    function fn() {
      new Action() // eslint-disable-line no-new
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a description', () => {

    function fn() {
      new Action({ // eslint-disable-line no-new
        name: 'example',
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have an input definition', () => {

    function fn() {
      new Action({ // eslint-disable-line no-new
        name: 'example',
        description: 'do something',
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a valid input definition', () => {

    function fn() {
      new Action({ // eslint-disable-line no-new
        name: 'example',
        description: 'do something',
        input: 123,
        output: 456
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have an output definition', () => {

    function fn() {
      new Action({ // eslint-disable-line no-new
        name: 'example',
        description: 'do something',
        input() {},
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a valid output definition', () => {

    function fn() {
      new Action({ // eslint-disable-line no-new
        name: 'example',
        description: 'do something',
        input() {},
        output: 'wrong'
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have valid input and output definitions when generated via a function', () => {

    const action = new Action({ // eslint-disable-line no-new
      name: 'example',
      description: 'do something',
      input() {
        return 123
      },
      output() {},
      resolve() {},
    })

    function fn1() {
      action.getInput()
    }

    function fn2() {
      action.getOutput()
    }

    expect(fn1).toThrowErrorMatchingSnapshot();
    expect(fn2).toThrowErrorMatchingSnapshot();

  })


  it('should have valid input data types', () => {

    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {
        someAttr: {}
      },
      output: {},
      resolve() {},
    })

    function fn() {
      action.getInput()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have valid output data types', () => {

    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {
        someAttr: {}
      },
      resolve() {},
    })

    function fn() {
      action.getOutput()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have valid input attributes if valid defaultValue functions', () => {

    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {
        someAttr: {
          type: DataTypeString,
          defaultValue: 'not-a-function'
        }
      },
      output: {},
      resolve() {},
    })

    function fn() {
      action.getInput()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a valid resolve function', () => {

    function fn() {
      new Action({ // eslint-disable-line no-new
        name: 'example',
        description: 'do something',
        input: {},
        output: {},
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should return it\'s name', () => {

    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {},
      resolve() {},
    })

    expect(action.name).toBe('example');
    expect(String(action)).toBe('example');

  })


  it('should return input and output params', () => {

    const action1 = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {},
      resolve() {},
    })

    const input1 = action1.getInput()
    const output1 = action1.getOutput()

    expect(input1).toEqual({});
    expect(output1).toEqual({});

    const hasInputParams1 = action1.hasInputParams()
    const hasOutputParams1 = action1.hasOutputParams()

    expect(hasInputParams1).toBe(false);
    expect(hasOutputParams1).toBe(false);


    const action2 = new Action({
      name: 'example',
      description: 'do something',
      input: {
        firstName: {
          type: DataTypeString,
          defaultValue() {
            return 'Waldo'
          }
        },
        lastName: {
          type: DataTypeString,
          required: true
        },
        about: {
          description: 'Just some description',
          type: buildObjectDataType({
            attributes: {
              favouriteActor: {
                type: DataTypeString,
                description: 'One more description'
              },
            }
          })
        },
      },
      output: {
        luckyNumber: {
          type: DataTypeString,
        },
      },
      resolve() {},
    })

    const input2 = action2.getInput()
    const output2 = action2.getOutput()

    expect(String(input2.firstName.type)).toBe('DataTypeString');
    expect(String(input2.lastName.type)).toBe('DataTypeString');
    expect(input2.firstName.required).toBe(false);
    expect(input2.lastName.required).toBe(true);
    expect(typeof input2.firstName.defaultValue).toBe('function');

    const nestedInput2 = input2.about.type.getAttributes()
    expect(String(nestedInput2.favouriteActor.type)).toBe('DataTypeString')

    expect(String(output2.luckyNumber.type)).toBe('DataTypeString');

    const hasInputParams2 = action2.hasInputParams()
    const hasOutputParams2 = action2.hasOutputParams()

    expect(hasInputParams2).toBe(true);
    expect(hasOutputParams2).toBe(true);

  })


  describe('isAction', () => {

    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {},
      resolve() {},
    })

    it('should recognize objects of type Action', () => {

      function fn() {
        passOrThrow(
          isAction(action),
          () => 'This error will never happen'
        )
      }

      expect(fn).not.toThrow()

    })


    it('should recognize non-Action objects', () => {

      function fn() {
        passOrThrow(
          isAction({}) ||
          isAction(function test() {}) ||
          isAction(Error),
          () => 'Not a Action object'
        )
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })

  })

})
