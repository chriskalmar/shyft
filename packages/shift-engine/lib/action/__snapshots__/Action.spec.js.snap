// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action isAction should recognize non-Action objects 1`] = `"Not a Action object"`;

exports[`Action should have a description 1`] = `"Missing description for action 'example'"`;

exports[`Action should have a name 1`] = `"Missing action name"`;

exports[`Action should have a valid input definition 1`] = `"Action 'example' needs an input definition as a map or a function returning such a map"`;

exports[`Action should have a valid output definition 1`] = `"Action 'example' needs an output definition as a map or a function returning such a map"`;

exports[`Action should have a valid resolve function 1`] = `"Action 'example' needs a resolve function"`;

exports[`Action should have an input definition 1`] = `"Missing input definition for action 'example'"`;

exports[`Action should have an output definition 1`] = `"Missing output definition for action 'example'"`;

exports[`Action should have valid input and output definitions when generated via a function 1`] = `"Input definition function for action 'example' does not return a map"`;

exports[`Action should have valid input and output definitions when generated via a function 2`] = `"Output definition function for action 'example' does not return a map"`;

exports[`Action should have valid input attributes if valid defaultValue functions 1`] = `"'example.someAttr' has an invalid defaultValue function'"`;

exports[`Action should have valid input data types 1`] = `"'example.someAttr' has invalid data type 'undefined'"`;

exports[`Action should have valid output data types 1`] = `"'example.someAttr' has invalid data type 'undefined'"`;
